File: java/HappyNumber.java
/*
 * Filename: HappyNumber.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 05 Jun 2015 06:43:54 AM CST
 */

public class Solution {
	public boolean isHappy(int n) {
		if (n == 1) {
			return true;
		}

		HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
		while (true) {
			map.put(n, n);
			n = ssd(n);
			if (map.containsKey(n)) {
				return false;
			} else if (n == 1) {
				return true;
			}
		}

		return false;
	}

	private int ssd(int n) {
		int res = 0;
		int digit;
		while (n > 0) {
			digit = n % 10;
			res += digit * digit;
			n /= 10;
		}

		return res;
	}
}

File: cpp/happy_number.cpp
/*
 * Filename: happy_number.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Mon 16 Nov 2015 11:14:23 PM EST
 * Description: 
 */

#include <iostream>
#include <map>

using namespace std;

class Solution {
	public:
		bool isHappy(int n) {
			map<int, int> map;

			while (true)
			{
				if (n == 1)
				{
					return true;
				}
				else if (map.find(n) != map.end())
				{
					return false;
				}
				else
				{
					map[n] = 1;
				}

				int ss = 0;
				while (n)
				{
					int r = n % 10;
					ss += r * r;
					n = n / 10;
				}

				n = ss;
			}
		}
};
